tests/unit_tests/verRctNonSemanticsSimple.cpp

```diff
@@ -0,0 +1,220 @@
+            (
+                cryptonote::parse_and_validate_tx_from_blob(tx_blob, transaction),
+                "TX blob could not be parsed"
+            );
+
+            const auto prefix_hash = cryptonote::get_transaction_prefix_hash(transaction);
+            const auto pubkeys = construct_original_input_pubkeys();
+
+            CHECK_AND_ASSERT_THROW_MES
+            (
+                cryptonote::Blockchain::expand_transaction_2(transaction, prefix_hash, pubkeys),
+                "expand_transactions_2 failed"
+            );
+
+            success = true;
+        });
+
+        CHECK_AND_ASSERT_THROW_MES(success, "bad one-time transaction deserialization/expansion");
+
+        return transaction;
+    }
+
+    const rct::rctSig& get_original_sig()
+    {
+        return get_original_transaction().rct_signatures;
+    }
+
+    const std::string& original_sig_hex()
+    {
+        static std::string hexstr;
+        static std::once_flag hexstr_of;
+
+        std::call_once(hexstr_of, [&]()
+        {
+            hexstr = rct_sig_to_bin_hex_string(get_original_sig());
+        });
+
+        return hexstr;
+    }
+} // anonymous namespace
+
+TEST(verRctNonSemanticsSimple, tx_preconditions)
+{
+    // If this unit test fails, something changed about transaction deserialization / expansion or
+    // something changed about RingCT signature verification.
+
+    const cryptonote::transaction& tx = get_original_transaction();
+    const rct::rctSig& rs = get_original_sig();
+
+    ASSERT_EQ(1, tx.vin.size());
+    const rct::key expected_sig_msg = rct::hash2rct(cryptonote::get_transaction_prefix_hash(tx));
+    ASSERT_EQ(expected_sig_msg, rs.message);
+    ASSERT_EQ(1, rs.mixRing.size());
+    ASSERT_EQ(16, rs.mixRing[0].size());
+    ASSERT_EQ(0, rs.pseudoOuts.size());
+
+    ASSERT_TRUE(rct::verRctSemanticsSimple(rs));
+    ASSERT_TRUE(rct::verRctNonSemanticsSimple(rs));
+    ASSERT_TRUE(rct::verRctSimple(rs));
+}
+
+TEST(verRctNonSemanticsSimple, cached_serialize_assumption)
+{
+    // Hello, future visitors. If this unit test fails, than an important, consensus-related
+    // assumption has been broken and rct::verRctNonSemanticsSimpleCached() in file
+    // src/ringct/rctSigs.{h/cpp} needs to be reassessed for validity. The assumption is below:
+
+    // For all values of S and S' where S and S' are of type rct::rctSig and S =/= S',
+    // ::do_serialize(S) =/= ::do_serialize(S') must hold.
+    // In other words, ::do_serialize() must create a unique representation of rct::rctSig
+
+    // These are the subtests most likely to fail. Fields 'message' and 'mixRing' are not serialized
+    // when sent over the wire, since they can be reconstructed from transaction data.
+    // How signatures are serialized for the blockchain can be found in the methods
+    // rct::rctSigBase::serialize_rctsig_base and rct::rctSigPrunable::serialize_rctsig_prunable.
+    CACHED_SERIALIZATION_ASSUMPTION_SUBTEST(message.bytes[31]++)
```

Comment:
Not too excited about using a macro, at least what you are doing is simple enough. The main alternative that springs to mind is using a lambda passed to a helper function that replaces this macro.

Reply:
One big benefit of using a macro here is that when/if the unit test fails, we get the line number of that specific instance of the test, instead of the same line number of the `EXPECT_NE` inside the helper function.

Reply:
What you can do is return `bool` on your test helper function, then `EXPECT_TRUE()` in your `TEST()`.

Reply:
This is true, but it gets more complicated with `EXPECT_EQ`, `EXPECT_NE`, etc...

Reply:
In those cases you just use an assert with clear message, then `EXPECT_NO_THROW()` in the test body.